Regression(r291141) Flashing when hovering photos on nytimes.com

cdumez opened the pull request:
#### 04e2b883c9aefdbe1b533b9bf256bde9cb1ddf29
<pre>
Regression(r291141) Flashing when hovering photos on nytimes.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=241921">https://bugs.webkit.org/show_bug.cgi?id=241921</a>
&lt;rdar://95098539&gt;

Reviewed by Geoffrey Garen.

Revert r291141 since it introduced the flashing of photos on nytimes.com and we suspect
it caused some crashes as well.

Instead, do a much more minimal fix in ImageLoader::updatedHasPendingEvent() to fix the
memory leak that r291141 was trying to fix. In updatedHasPendingEvent() we no longer
protect the element if the load is deferred indefinitely (which is the case for lazy
image loading). In the lazy image loading case, when the load actually starts,
ImageLoader::updateFromElement() gets called again, which calls updatedHasPendingEvent()
which will protect the element then.

No new tests, covered by fast/dom/lazy-image-loading-document-leak.html that is still
passing.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::ImageLoader):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::didUpdateCachedImage):
(WebCore::ImageLoader::updatedHasPendingEvent):
(WebCore::ImageLoader::didStartLoading): Deleted.
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::load):
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::didStartLoading): Deleted.

Canonical link: <a href="https://commits.webkit.org/251802@main">https://commits.webkit.org/251802@main</a>
</pre>

geoffreygaren commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -448,7 +432,12 @@ void ImageLoader::updatedHasPendingEvent()
+    // protect it once the load has actually started.
+    bool imageWillBeLoadedLater = m_hasPendingLoadEvent && m_image && !m_image->isLoading() && m_image->stillNeedsLoad();
+
+    m_elementIsProtected = (m_hasPendingLoadEvent && !imageWillBeLoadedLater) || m_hasPendingErrorEvent;
```
Seems like including m_hasPendingLoadEvent in imageWillBeLoadedLater is redundant, no?

cdumez replied:
Yes, will fix.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Regression(r291141) Flashing when hovering photos on nytimes.com
https://bugs.webkit.org/show_bug.cgi?id=241921
<rdar://95098539>

Reviewed by Geoffrey Garen.

Revert r291141 since it introduced the flashing of photos on nytimes.com and we suspect
it caused some crashes as well.

Instead, do a much more minimal fix in ImageLoader::updatedHasPendingEvent() to fix the
memory leak that r291141 was trying to fix. In updatedHasPendingEvent() we no longer
protect the element if the load is deferred indefinitely (which is the case for lazy
image loading). In the lazy image loading case, when the load actually starts,
ImageLoader::updateFromElement() gets called again, which calls updatedHasPendingEvent()
which will protect the element then.

No new tests, covered by fast/dom/lazy-image-loading-document-leak.html that is still
passing.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::ImageLoader):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::didUpdateCachedImage):
(WebCore::ImageLoader::updatedHasPendingEvent):
(WebCore::ImageLoader::didStartLoading): Deleted.
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::load):
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::didStartLoading): Deleted.

Canonical link: https://commits.webkit.org/251802@main
